File: app/graphics/draw-expression.ts
/** Revision 27 expression bytecode. ELAPSED is a bridge binding, never sent to firmware. */
export const ExprOp = {
  I32: 1, F32: 2, DUP: 3, DROP: 4, SWAP: 5,
  IADD: 16, ISUB: 17, IMUL: 18, IDIV: 19, IMOD: 20, INEG: 21, IMIN: 22, IMAX: 23,
  FADD: 32, FSUB: 33, FMUL: 34, FDIV: 35, FNEG: 36, FMIN: 37, FMAX: 38,
  I2F: 48, F2I: 49, TIME: 50, LERP: 64, SMOOTHSTEP: 65,
  EASE_IN_QUAD: 66, EASE_OUT_QUAD: 67, EASE_IN_OUT_QUAD: 68,
  EASE_IN_CUBIC: 69, EASE_OUT_CUBIC: 70, EASE_IN_OUT_CUBIC: 71,
  ELAPSED: 128,
} as const;

export function integer(value: number, min: number, max: number): number {
  if (!Number.isInteger(value) || value < min || value > max) throw new Error('Draw integer out of range');
  return value;
}
export function varint(value: number, signed = true): number[] {
  integer(value, signed ? -2147483648 : 0, signed ? 2147483647 : 4294967295);
  const widths = [7, 14, 21, 28, 32], prefixes = [0, 128, 192, 224, 240];
  const size = widths.findIndex(bits => bits === 32 || (signed
    ? value >= -(2 ** (bits - 1)) && value < 2 ** (bits - 1) : value < 2 ** bits)) + 1;
  const bytes = [size === 5 ? 240 : prefixes[size - 1] | ((value >>> (8 * (size - 1))) & ((1 << (8 - size)) - 1))];
  for (let i = 1; i < size; i++) bytes.push((value >>> (8 * (size - 1 - i))) & 255);
  return bytes;
}

export class DrawBytes {
  offset = 0;
  constructor(readonly bytes: Uint8Array) {}
  take(size: number): Uint8Array {
    integer(size, 0, this.bytes.length - this.offset);
    const result = this.bytes.slice(this.offset, this.offset + size); this.offset += size; return result;
  }
  u8(): number { return this.take(1)[0]; }
  u16(): number { return this.u8() | this.u8() << 8; }
  i16(): number { return this.u16() << 16 >> 16; }
  u32(): number { return (this.u16() | this.u16() << 16) >>> 0; }
  variable(signed = true): number {
    const first = this.u8();
    const size = first < 128 ? 1 : first < 192 ? 2 : first < 224 ? 3 : first < 240 ? 4 : first === 240 ? 5 : 0;
    if (!size) throw new Error('Reserved extended-varint prefix');
    let value = size === 5 ? 0 : first & ((1 << (8 - size)) - 1);
    for (let i = 1; i < size; i++) value = (value << 8) | this.u8();
    const bits = [7, 14, 21, 28, 32][size - 1];
    return signed ? bits < 32 ? value << (32 - bits) >> (32 - bits) : value : value >>> 0;
  }
}

/** Immutable, typed expression builder; operations use firmware i32/f32 semantics. */
export class DrawExpression<T extends 'i32' | 'f32'> {
  private constructor(readonly type: T, readonly code: readonly number[]) {
    // Leave room for native elapsed bindings and surface translation.
    if (code.length > 900) throw new Error('Draw expression too large');
  }
  static i32(value: number): DrawExpression<'i32'> { return new DrawExpression('i32', [ExprOp.I32, ...varint(value)]); }
  static f32(value: number): DrawExpression<'f32'> {
    if (!Number.isFinite(Math.fround(value))) throw new Error('Invalid draw float');
    const bytes = new Uint8Array(4); new DataView(bytes.buffer).setFloat32(0, value, true);
    return new DrawExpression('f32', [ExprOp.F32, ...bytes]);
  }
  /** Time already elapsed on this list's timeline when the next PRESENT happens. */
  static elapsed(): DrawExpression<'i32'> { return new DrawExpression('i32', [ExprOp.ELAPSED]); }
  /**
   * A 0..1 timeline fraction that survives subsequent PRESENT clock resets.
   * With a delay, it stays 0 for the first delayMs of the timeline; a
   * negative delay continues a motion that started before the timeline.
   */
  static progress(durationMs: number, delayMs = 0): DrawExpression<'f32'> {
    integer(delayMs, -2147483647, 2147483647);
    if (delayMs + durationMs <= 0) return DrawExpression.f32(1);
    const duration = DrawExpression.i32(integer(durationMs, 1, 2147483647));
    const end = DrawExpression.i32(integer(delayMs + durationMs, 1, 2147483647));
    const elapsed = DrawExpression.elapsed().min(end);
    let total = end.sub(elapsed).max(DrawExpression.i32(0)).time().add(elapsed);
    if (delayMs) total = total.sub(DrawExpression.i32(delayMs)).max(DrawExpression.i32(0));
    return total.toFloat().div(duration.toFloat());
  }
  static decode(code: Uint8Array): DrawExpression<'i32'> { return new DrawExpression('i32', Array.from(code)); }
  private binary(other: DrawExpression<T>, i: number, f: number): DrawExpression<T> {
    if (other.type !== this.type) throw new Error('Draw expression type mismatch');
    return new DrawExpression(this.type, [...this.code, ...other.code, this.type === 'i32' ? i : f]);
  }
  add(b: DrawExpression<T>): DrawExpression<T> { return this.binary(b, ExprOp.IADD, ExprOp.FADD); }
  sub(b: DrawExpression<T>): DrawExpression<T> { return this.binary(b, ExprOp.ISUB, ExprOp.FSUB); }
  mul(b: DrawExpression<T>): DrawExpression<T> { return this.binary(b, ExprOp.IMUL, ExprOp.FMUL); }
  div(b: DrawExpression<T>): DrawExpression<T> { return this.binary(b, ExprOp.IDIV, ExprOp.FDIV); }
  min(b: DrawExpression<T>): DrawExpression<T> { return this.binary(b, ExprOp.IMIN, ExprOp.FMIN); }
  max(b: DrawExpression<T>): DrawExpression<T> { return this.binary(b, ExprOp.IMAX, ExprOp.FMAX); }
  mod(this: DrawExpression<'i32'>, b: DrawExpression<'i32'>): DrawExpression<'i32'> { return this.binary(b, ExprOp.IMOD, ExprOp.IMOD); }
  neg(): DrawExpression<T> { return new DrawExpression(this.type, [...this.code, this.type === 'i32' ? ExprOp.INEG : ExprOp.FNEG]); }
  toFloat(this: DrawExpression<'i32'>): DrawExpression<'f32'> { return new DrawExpression('f32', [...this.code, ExprOp.I2F]); }
  toInt(this: DrawExpression<'f32'>): DrawExpression<'i32'> { return new DrawExpression('i32', [...this.code, ExprOp.F2I]); }
  time(this: DrawExpression<'i32'>): DrawExpression<'i32'> { return new DrawExpression('i32', [...this.code, ExprOp.TIME]); }
  ease(this: DrawExpression<'f32'>, op: number = ExprOp.SMOOTHSTEP): DrawExpression<'f32'> {
    integer(op, ExprOp.SMOOTHSTEP, ExprOp.EASE_IN_OUT_CUBIC);
    return new DrawExpression('f32', [...this.code, op]);
  }
  lerp(this: DrawExpression<'f32'>, from: DrawExpression<'f32'>, to: DrawExpression<'f32'>): DrawExpression<'f32'> {
    return new DrawExpression('f32', [...from.code, ...to.code, ...this.code, ExprOp.LERP]);
  }
}
export type DrawValue = number | DrawExpression<'i32'>;
export function encodeValue(value: DrawValue): number[] {
  return typeof value === 'number' ? varint(value) : [255, ...varint(value.code.length, false), ...value.code];
}
export function readValue(reader: DrawBytes): DrawValue {
  if (reader.bytes[reader.offset] !== 255) return reader.variable();
  reader.u8(); return DrawExpression.decode(reader.take(reader.variable(false)));
}

/** Software preview of the same VM; invalid programs return zero and discard pending animation. */
export function evaluate(value: DrawValue, elapsed: number, presentTime = 0): { value: number; pending: boolean } {
  if (typeof value === 'number') return { value, pending: false };
  const reader = new DrawBytes(Uint8Array.from(value.code)), stack: { value: number; float: boolean }[] = [];
  let pending = false;
  const push = (value: number, float = false) => {
    value = float ? Math.fround(value) : value | 0;
    if (!Number.isFinite(value) || stack.length === 32) throw new Error('Invalid expression stack');
    stack.push({ value, float });
  };
  const pop = (float: boolean) => { const v = stack.pop(); if (!v || v.float !== float) throw new Error('Expression type mismatch'); return v.value; };
  const toInt = (v: number) => { if (v < -2147483648 || v >= 2147483648) throw new Error('Invalid integer conversion'); return Math.trunc(v); };
  try {
    while (reader.offset < reader.bytes.length) {
      const op = reader.u8();
      if (op === ExprOp.I32) push(reader.variable());
      else if (op === ExprOp.F32) { const b = reader.take(4); push(new DataView(b.buffer).getFloat32(0, true), true); }
      else if (op === ExprOp.ELAPSED) push(Math.min(2147483647, Math.max(0, elapsed)));
      else if (op === ExprOp.I2F) push(pop(false), true);
      else if (op === ExprOp.F2I) push(toInt(pop(true)));
      else if (op === ExprOp.TIME) { const max = pop(false); if (max < 0) throw new Error('Invalid time'); const t = Math.min(Math.max(0, presentTime), max); pending ||= t !== max; push(t); }
      else if (op === ExprOp.INEG || op === ExprOp.FNEG) push(-pop(op === ExprOp.FNEG), op === ExprOp.FNEG);
      else if (op >= ExprOp.IADD && op <= ExprOp.IMAX || op >= ExprOp.FADD && op <= ExprOp.FMAX) {
        const float = op >= ExprOp.FADD, b = pop(float), a = pop(float);
        const key = float ? op - ExprOp.FADD : op - ExprOp.IADD;
        if ((key === 3 || !float && key === 4) && b === 0) throw new Error('Division by zero');
        const result = key === 0 ? a + b : key === 1 ? a - b : key === 2 ? float ? a * b : Math.imul(a, b) : key === 3 ? a / b : !float && key === 4 ? a % b : key === (float ? 5 : 6) ? Math.min(a, b) : Math.max(a, b);
        push(result, float);
      } else if (op === ExprOp.LERP) { const t = pop(true), b = pop(true), a = pop(true); push(a + Math.fround(Math.fround(b - a) * t), true); }
      else if (op >= ExprOp.SMOOTHSTEP && op <= ExprOp.EASE_IN_OUT_CUBIC) {
        const t = Math.max(0, Math.min(1, pop(true))), u = Math.fround(1 - t), f = Math.fround;
        let result: number;
        switch (op) {
          case ExprOp.SMOOTHSTEP: result = f(t * t) * f(3 - f(2 * t)); break;
          case ExprOp.EASE_IN_QUAD: result = t * t; break;
          case ExprOp.EASE_OUT_QUAD: result = 1 - f(u * u); break;
          case ExprOp.EASE_IN_OUT_QUAD: result = t < .5 ? f(2 * t) * t : 1 - f(f(2 * u) * u); break;
          case ExprOp.EASE_IN_CUBIC: result = f(t * t) * t; break;
          case ExprOp.EASE_OUT_CUBIC: result = 1 - f(f(u * u) * u); break;
          default: result = t < .5 ? f(f(4 * t) * t) * t : 1 - f(f(f(4 * u) * u) * u);
        }
        push(result, true);
      } else if (op === ExprOp.DUP) { const a = stack[stack.length - 1]; if (!a) throw new Error('Underflow'); push(a.value, a.float); }
      else if (op === ExprOp.DROP) { if (!stack.pop()) throw new Error('Underflow'); }
      else if (op === ExprOp.SWAP) { if (stack.length < 2) throw new Error('Underflow'); const a = stack.pop()!, b = stack.pop()!; stack.push(a, b); }
      else throw new Error('Unknown expression opcode');
    }
    const top = stack.pop(); if (!top) throw new Error('Empty expression');
    return { value: top.float ? toInt(top.value) : top.value, pending };
  } catch { return { value: 0, pending: false }; }
}

File: native/kotlin/shared/src/commonMain/kotlin/com/faceclaw/app/g2protocol/DrawExpression.kt
package com.faceclaw.app

/** Literal or branchless expression used by rounded-rectangle x/y in revision 27. */
sealed class DrawValue {
    class Integer(val value: Int) : DrawValue()
    class Expression(val program: ByteArray) : DrawValue()

    companion object {
        /** Smoothstep interpolation; time is measured from the next PRESENT. */
        fun animate(from: Int, to: Int, durationMs: Int, elapsedMs: Int = 0): DrawValue {
            require(durationMs > 0 && elapsedMs >= 0)
            if (from == to || elapsedMs >= durationMs) return Integer(to)
            val code = ByteSink()
            fun integer(value: Int) {
                code.write(DrawExpression.PUSH_I32)
                code.write(ExtendedVarint.signed(value))
            }
            integer(from)
            code.write(DrawExpression.I2F)
            integer(to)
            code.write(DrawExpression.I2F)
            integer(durationMs - elapsedMs)
            code.write(DrawExpression.TIME)
            integer(elapsedMs)
            code.write(DrawExpression.IADD)
            code.write(DrawExpression.I2F)
            integer(durationMs)
            code.write(DrawExpression.I2F)
            code.write(DrawExpression.FDIV)
            code.write(DrawExpression.SMOOTHSTEP)
            code.write(DrawExpression.LERP)
            code.write(DrawExpression.F2I)
            return Expression(code.toByteArray())
        }
    }
}

/** Prefix payload is most-significant first; continuation bytes use all eight bits. */
internal object ExtendedVarint {
    private val widths = intArrayOf(7, 14, 21, 28, 32)
    private val prefixes = intArrayOf(0, 0x80, 0xc0, 0xe0, 0xf0)

    fun signed(value: Int): ByteArray {
        val size = widths.indexOfFirst { bits ->
            bits == 32 || value.toLong() in -(1L shl (bits - 1)) until (1L shl (bits - 1))
        } + 1
        return encode(value, size)
    }

    fun unsigned(value: UInt): ByteArray {
        val size = widths.indexOfFirst { it == 32 || value.toLong() < (1L shl it) } + 1
        return encode(value.toInt(), size)
    }

    private fun encode(value: Int, size: Int): ByteArray {
        val bytes = ByteArray(size)
        if (size == 5) {
            bytes[0] = 0xf0.toByte()
        } else {
            val mask = (1 shl (8 - size)) - 1
            bytes[0] = (prefixes[size - 1] or ((value ushr (8 * (size - 1))) and mask)).toByte()
        }
        for (i in 1 until size) bytes[i] = (value ushr (8 * (size - 1 - i))).toByte()
        return bytes
    }

    fun read(reader: DrawReader, signed: Boolean, first: Int = reader.readU8()): Int {
        val size = when {
            first < 0x80 -> 1
            first < 0xc0 -> 2
            first < 0xe0 -> 3
            first < 0xf0 -> 4
            first == 0xf0 -> 5
            else -> error("Reserved extended-varint prefix")
        }
        var value = if (size == 5) 0 else first and ((1 shl (8 - size)) - 1)
        repeat(size - 1) { value = (value shl 8) or reader.readU8() }
        val bits = widths[size - 1]
        return if (signed && bits < 32) (value shl (32 - bits)) shr (32 - bits) else value
    }

    fun write(value: DrawValue): ByteArray = when (value) {
        is DrawValue.Integer -> signed(value.value)
        is DrawValue.Expression -> {
            require(value.program.size <= DrawExpression.MAX_BYTES)
            byteArrayOf(0xff.toByte()) + unsigned(value.program.size.toUInt()) + value.program
        }
    }

    fun evaluate(reader: DrawReader, frame: DrawEvaluation): Int {
        val first = reader.readU8()
        if (first != 0xff) return read(reader, true, first)
        val length = read(reader, false).toUInt()
        // Outer framing errors reject the call; errors inside a framed program yield zero.
        require(length <= reader.remaining.toUInt())
        val program = reader.readSlice(length.toInt())
        if (length > DrawExpression.MAX_BYTES.toUInt()) return 0
        return DrawExpression.evaluate(program, frame)
    }
}

internal class DrawEvaluation(val elapsedMs: Long) {
    var animationPending = false
}

/** Typed i32/f32 stack machine. Keep opcode IDs and semantics in sync with draw_expression.c. */
internal object DrawExpression {
    const val MAX_BYTES = 1024
    const val MAX_STACK = 32
    const val PUSH_I32 = 1
    const val PUSH_F32 = 2
    const val DUP = 3
    const val DROP = 4
    const val SWAP = 5
    const val IADD = 16
    const val ISUB = 17
    const val IMUL = 18
    const val IDIV = 19
    const val IMOD = 20
    const val INEG = 21
    const val IMIN = 22
    const val IMAX = 23
    const val FADD = 32
    const val FSUB = 33
    const val FMUL = 34
    const val FDIV = 35
    const val FNEG = 36
    const val FMIN = 37
    const val FMAX = 38
    const val I2F = 48
    const val F2I = 49
    const val TIME = 50
    const val LERP = 64
    const val SMOOTHSTEP = 65
    const val EASE_IN_QUAD = 66
    const val EASE_OUT_QUAD = 67
    const val EASE_IN_OUT_QUAD = 68
    const val EASE_IN_CUBIC = 69
    const val EASE_OUT_CUBIC = 70
    const val EASE_IN_OUT_CUBIC = 71

    fun evaluate(reader: DrawReader, frame: DrawEvaluation): Int {
        val bits = IntArray(MAX_STACK)
        val floats = BooleanArray(MAX_STACK)
        var size = 0
        var pending = false
        fun push(value: Int, float: Boolean = false) {
            require(size < MAX_STACK)
            bits[size] = value
            floats[size++] = float
        }
        fun integer(): Int {
            require(size > 0 && !floats[size - 1])
            return bits[--size]
        }
        fun float(): Float {
            require(size > 0 && floats[size - 1])
            return Float.fromBits(bits[--size])
        }
        fun pushFloat(value: Float) {
            require(value.isFinite())
            push(value.toBits(), true)
        }
        fun toInteger(value: Float): Int {
            require(value.isFinite() && value >= -2147483648f && value < 2147483648f)
            return value.toInt()
        }
        try {
            require(reader.remaining <= MAX_BYTES)
            while (reader.remaining > 0) {
                when (val op = reader.readU8()) {
                    PUSH_I32 -> push(ExtendedVarint.read(reader, true))
                    PUSH_F32 -> {
                        val value = reader.readU8() or (reader.readU8() shl 8) or
                            (reader.readU8() shl 16) or (reader.readU8() shl 24)
                        pushFloat(Float.fromBits(value))
                    }
                    DUP -> {
                        require(size > 0)
                        push(bits[size - 1], floats[size - 1])
                    }
                    DROP -> { require(size > 0); size-- }
                    SWAP -> {
                        require(size >= 2)
                        val value = bits[size - 1]
                        val type = floats[size - 1]
                        bits[size - 1] = bits[size - 2]
                        floats[size - 1] = floats[size - 2]
                        bits[size - 2] = value
                        floats[size - 2] = type
                    }
                    IADD, ISUB, IMUL, IDIV, IMOD, IMIN, IMAX -> {
                        val b = integer()
                        val a = integer()
                        push(when (op) {
                            IADD -> a + b
                            ISUB -> a - b
                            IMUL -> a * b
                            IDIV -> { require(b != 0); if (a == Int.MIN_VALUE && b == -1) a else a / b }
                            IMOD -> { require(b != 0); if (a == Int.MIN_VALUE && b == -1) 0 else a % b }
                            IMIN -> minOf(a, b)
                            else -> maxOf(a, b)
                        })
                    }
                    INEG -> push(-integer())
                    FADD, FSUB, FMUL, FDIV, FMIN, FMAX -> {
                        val b = float()
                        val a = float()
                        pushFloat(when (op) {
                            FADD -> a + b
                            FSUB -> a - b
                            FMUL -> a * b
                            FDIV -> { require(b != 0f); a / b }
                            FMIN -> minOf(a, b)
                            else -> maxOf(a, b)
                        })
                    }
                    FNEG -> pushFloat(-float())
                    I2F -> pushFloat(integer().toFloat())
                    F2I -> push(toInteger(float()))
                    TIME -> {
                        val maximum = integer()
                        require(maximum >= 0)
                        val time = frame.elapsedMs.coerceIn(0, maximum.toLong()).toInt()
                        if (time != maximum) pending = true
                        push(time)
                    }
                    LERP -> {
                        val t = float()
                        val b = float()
                        val a = float()
                        pushFloat(a + (b - a) * t)
                    }
                    in SMOOTHSTEP..EASE_IN_OUT_CUBIC -> {
                        val t = float().coerceIn(0f, 1f)
                        val u = 1f - t
                        pushFloat(when (op) {
                            SMOOTHSTEP -> t * t * (3f - 2f * t)
                            EASE_IN_QUAD -> t * t
                            EASE_OUT_QUAD -> 1f - u * u
                            EASE_IN_OUT_QUAD -> if (t < .5f) 2f * t * t else 1f - 2f * u * u
                            EASE_IN_CUBIC -> t * t * t
                            EASE_OUT_CUBIC -> 1f - u * u * u
                            else -> if (t < .5f) 4f * t * t * t else 1f - 4f * u * u * u
                        })
                    }
                    else -> error("Unknown expression opcode")
                }
            }
            require(size > 0)
            val result = if (floats[size - 1]) toInteger(float()) else integer()
            frame.animationPending = frame.animationPending || pending
            return result
        } catch (_: IllegalArgumentException) {
            return 0
        } catch (_: IllegalStateException) {
            return 0
        }
    }
}
